presto-hive/src/main/java/io/prestosql/plugin/hive/metastore/thrift/ThriftHiveMetastore.java

```diff
@@ -622,6 +636,56 @@ private void deletePartitionColumnStatistics(HiveIdentity identity, String datab
         }
     }
 
+    private void setColumnStatistics(HiveIdentity identity, String objectName, List<ColumnStatisticsObj> statistics, Call1<List<ColumnStatisticsObj>> saveColumnStatistics)
+            throws TException
+    {
+        boolean containsDateStatistics = statistics.stream().anyMatch(stats -> stats.getStatsData().isSetDateStats());
+
+        Optional<Boolean> metastoreSupportsDateStatistics = this.metastoreSupportsDateStatistics.get();
+        if (containsDateStatistics && metastoreSupportsDateStatistics.equals(Optional.of(FALSE))) {
+            log.debug("Skipping date statistics for %s because metastore does not support them", objectName);
+            statistics = statistics.stream()
+                    .filter(stats -> !stats.getStatsData().isSetDateStats())
+                    .collect(toImmutableList());
+            containsDateStatistics = false;
+        }
+
+        if (!containsDateStatistics || metastoreSupportsDateStatistics.equals(Optional.of(TRUE))) {
+            try (ThriftMetastoreClient client = createMetastoreClient(identity)) {
+                saveColumnStatistics.call(client, statistics);
+            }
+            return;
+        }
+
+        List<ColumnStatisticsObj> statisticsExceptDate = statistics.stream()
+                .filter(stats -> !stats.getStatsData().isSetDateStats())
+                .collect(toImmutableList());
+
+        List<ColumnStatisticsObj> dateStatistics = statistics.stream()
+                .filter(stats -> stats.getStatsData().isSetDateStats())
+                .collect(toImmutableList());
+
+        verify(!dateStatistics.isEmpty() && metastoreSupportsDateStatistics.equals(Optional.empty()));
+
+        try (ThriftMetastoreClient client = createMetastoreClient(identity)) {
+            saveColumnStatistics.call(client, statisticsExceptDate);
+
+            try {
+                saveColumnStatistics.call(client, dateStatistics);
+            }
+            catch (TException e) {
+                // When `dateStatistics.size() > 1` we expect something like "TApplicationException: Required field 'colName' is unset! Struct:ColumnStatisticsObj(colName:null, colType:null, statsData:null)"
+                // When `dateStatistics.size() == 1` we expect something like "TTransportException: java.net.SocketTimeoutException: Read timed out"
+                log.warn(e, "Failed to save date statistics for %s. Metastore might not support date statistics", objectName);
+                if (!statisticsExceptDate.isEmpty() && metastoreSetDateStatisticsFailures.incrementAndGet() >= MAX_SET_DATE_STATISTICS_ATTEMPTS) {
```

Comment:
Given:
```
    private static final int MAX_SET_DATE_STATISTICS_ATTEMPTS = 100;
 ... 
    private final CoalescingCounter metastoreSetDateStatisticsFailures = new CoalescingCounter(new Duration(1, SECONDS));
```
When:
```
 metastoreSetDateStatisticsFailures.incrementAndGet() >= MAX_SET_DATE_STATISTICS_ATTEMPTS
```

Does it require to call `metastoreSetDateStatisticsFailures.incrementAndGet()` 100*1000 times to make above condition evaluate to `true`?

Did you forget to measure how long actual call to HMS lasts and increment `metastoreSetDateStatisticsFailures` with that?

Reply:
it takes 100 calls, just they need to be set apart.
I don't want to conclude "metastore doesn't support date" due to some unrelated transient error.
Hence i am not counting failures than occur within proximity of each other